shell.php

```diff
@@ -241,6 +257,52 @@ function _onShellCmdKeyDown(event) {
                             eShellCmdInput.value = commandHistory[historyPosition];
                         }
                         break;
+                    case 'Tab':
+                        event.preventDefault();
+                        var currentCmd = eShellCmdInput.value.split(' ');
+                        if (currentCmd.length > 0) {
+                            if (currentCmd.length === 1) {
```

Comment:
This can be refactored like this (I put this in a function, taking into account the two previous comments):

```javascript
function featureHint() {
    if (eShellCmdInput.value.trim().length === 0) return;  // field is empty -> nothing to complete

    function _requestCallback(data) {
        if (data.files.length <= 1) return;  // no completion

        if (data.files.length === 2) {
            eShellCmdInput.value = data.files[0];
        } else {
            _insertCommand(eShellCmdInput.value);
            _insertStdout(data.files.join("\n"));
        }
    }

    var currentCmd = eShellCmdInput.value.split(" ");
    var type = (currentCmd.length === 1) ? "cmd" : "file";
    var fileName = (type === "cmd") ? currentCmd[0] : currentCmd[currentCmd.length - 1];

    makeRequest(
        "?feature=hint",
        {
            filename: fileName,
            cwd: CWD,
            type: type
        },
        _requestCallback
    );

}

```